README.md
# Kali-Updater


### INSTALL

- Download

    - `git clone https://github.com/Cashiuus/Kali-Updater.git kaliupdater`
    - `cd kaliupdater`
    - `cp default.py settings.py`

- Modify "settings.py" with desired configuration settings
- Run script (no arguments necessary at this time)
	`python kaliupdater.py`

### SUPPORT

Tested on versions:
- 2.7.9


Follow me on Twitter: http://twitter.com/Cashiuus


### LICENSE

    Copyright 2014 Jason Hoerner

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.



ChangeLog
A complete changelog can be found at: https://github.com/Cashiuus/

~~~~~~~~~~~~~~~~~~~
Date    2015-01-04
Version 1.5
~~~~~~~~~~~~~~~~~~~
    - Converted GIT_BASE_DIR to a list and moved to settings.py.
    - Added ability to auto-scan your git folders and auto-update them.
      We'll ignore catching errors for now to avoid screwing up a repo.
    - Added debug functionality and text coloring


~~~~~~~~~~~~~~~~~~~
Date    2014-12-29
Version 1.4
~~~~~~~~~~~~~~~~~~~
    - Built helper file "default.py" and migrated user-configured settings
    - Incorporated file/folder backup functionality with exclusions list
    - Improved Google Chrome installer/updater


~~~~~~~~~~~~~~~~~~~
Date    2014-03-20
Version 1.3
~~~~~~~~~~~~~~~~~~~
    - Changed git repository settings to nested mappings
    - Added support for special git projects that require unique configurations


~~~~~~~~~~~~~~~~~~~
Date    2014-03-20
Version 1.2
~~~~~~~~~~~~~~~~~~~
    - Added printout of current interpreter engine versions for reference


~~~~~~~~~~~~~~~~~~~
Date    2014-03-20
Version 1.1
~~~~~~~~~~~~~~~~~~~
    - Changed git clones list to a dictionary with each Git URL as value
    - Added ability to setup new git clone when dict item doesn't exist
    - Added option to add "Kali Bleeding Edge" repo to sources.list file
    - Fixed "raw_input()" to work in both Python 2.x and 3.x


~~~~~~~~~~~~~~~~~~~
Date    2014-03-20
Version 1.0
~~~~~~~~~~~~~~~~~~~
    - Drafted initial application


default.py
#!/usr/bin/env python
#
# default.py
#
# If you make changes to this file, rename it to "settings.py"

import os
# DEBUGGING
DO_DEBUG = True
BACKUP_PATH = os.path.join(os.path.expanduser('~'), 'Backups')
usr = os.path.expanduser('~')

# Want to fix portmapper issue at boot?
fixportmapper   = False
# This Enables git config option for 'color.ui'
gitcolorize     = True
# Backup user-specified files to an archive file
DO_BACKUPS      = True
# Setup system for Google Chrome; if installed it will update
DO_CHROME       = True
# Update all known Git repositories with 'git pull'
DO_GIT_REPOS    = False
# Update the Veil Framework using its own script
UPDATE_VEIL     = True

# List here folders that you are using for all git repos
# YOUR PRIMARY GIT DIRECTORY SHOULD BE FIRST. This is where
# your git_tools repositories will be cloned if they don't exist
GIT_BASE_DIRS = [
    os.path.expanduser('~/git'),
]

# Dotfiles are backed up, so these are exclusions
# This list can also include any files that may exist in
# folders you specify in the BACKUP_FILES section you wish to skip
EXCLUDE_FILES = [
    '.ICEauthority',
    '.pulse-cookie',
    '.rnd',
    '.rvmrc',
    '.xsession-errors',
    '.xsession-errors.old',
    '.zlogin',
    '.zshrc',
    ]


BACKUP_FILES = [
    usr + '/.local/',
    ]


kaliUpdater.py
#!/usr/bin/env python
#
#	Target Release:		Python 2.x
#	Version:			1.3
#	Updated:			12/04/2014
#	Created by: 		cashiuus@gmail.com
#
# Functionality:
#   1. Add Bleeding Edge Repository
#   2. Update & Upgrade all Kali apps using 'apt-get'
#   3. Update defined list of Git Clones; if nonexistant you can add them
# ====================================================================

from __future__ import print_function
import os
import platform
import subprocess
import sys
import time

# ----------------------------- #
#    Set configurable settings  #
# ----------------------------- #
tdelay 			= 2				# Delay script for network latency
fixportmapper 	= False			# Want to fix portmapper issue at boot?
gitcolorize		= True			# This Enables git config option for 'color.ui'
DO_CHROME		= False			# Setup system for Google Chrome
UPDATE_VEIL		= True			# Update the Veil framework using its own script
GIT_BASE_DIR	= os.path.expanduser('~/git')

#    List your existing GIT CLONES here
#    TODO: Get existing git clones automatically; only known way is too slow
#			If I can find them in background during core update, would be ideal
normal_git_tools = {
	'artillery':'https://github.com/trustedsec/artillery',
	'creds':'https://github.com/DanMcInerney/creds.py',
	'kaliupdater':'https://github.com/Cashiuus/kaliUpdater',
	'lair':'https://github.com/fishnetsecurity/Lair',
	'powersploit':'https://github.com/mattifestation/PowerSploit',
	'seclists':'https://github.com/danielmiessler/SecLists',
	'veil':'https://github.com/Veil-Framework/Veil',
	'vfeed':'https://github.com/toolswatch/vFeed',
}
special_git_tools = {
	# These projects prefer to be placed in the "/opt" directory instead
	# local_name, Dict value=(Tuple: git_url, installer_file)
	'smbexec':('https://github.com/pentestgeek/smbexec','install.sh'),
}

# Setup Python 2 & 3 'raw_input' compatibility
#
try:
	input = raw_input
except NameError:
	pass

# ----------------------------- #
#              BEGIN            #
# ----------------------------- #

# Check - Root user
# TODO: If not root, run with sudo
def root_check():
	if not (os.geteuid() == 0):
		print("[-] Not currently root user. Please fix.")
		exit(1)
	return

def make_dirs(path):
	if not os.path.isdir(path):
		os.makedirs(path)
	return

#    Update Kali core distro using Aptitude
#
def core_update():
	print("[+] Now updating Kali and Packages...")
	try:
		os.system("apt-get -qq update && apt-get -qq -y dist-upgrade && apt-get -y autoclean")
		os.system("updatedb")
		print("[+] Successfully updated Kali...moving along...")
	except:
		print("[-] Error attempting to update Kali. Please try again later.")

	time.sleep(tdelay)
	return


# -----------------------------
# Checking APT Repository List file
#
def bleedingRepoCheck():
	bleeding = 'deb http://repo.kali.org/kali kali-bleeding-edge main'
	repoFile = open('/etc/apt/sources.list', 'r')

	for line in repoFile.readlines():
		if bleeding == line.rstrip():
			check = 1
			break
		else:
			check = 0
	if check == 0:
		addrepo = input("\n[*] Would you like to add the Kali Bleeding Edge Repo? [y,N]: ")
		if (addrepo == 'y'):
			os.system("echo deb http://repo.kali.org/kali kali-bleeding-edge main >> /etc/apt/sources.list")
	return



# TODO: Add code to compare current versions with those found present
# Get installed version of supporting programming applications with dpkg
def get_versions():
	pVersion = str(sys.version_info[0]) + '.' + str(sys.version_info[1]) + '.' + str(sys.version_info[2])
	print("\t[*] Python Version:\t{}".format(pVersion))

	try:
		out_bytes = subprocess.check_output(['ruby', '-v'])
		out_text = out_bytes.decode('UTF-8')
		out_text = out_text.split(' ')
		rVersion = out_text[1]
		print("\t[*] Ruby Version:\t", rVersion, sep='')
	except:
		print("\t[*] Ruby Version:\t Not Installed", sep='')

	try:
		out_bytes = subprocess.check_output(['gem', '-v'])
		gVersion = out_bytes.decode('UTF-8')
		gVersion = gVersion.rstrip()
		print("\t[*] Gem Version:\t", gVersion, sep='')
	except:
		print("\t[*] Gem Version:\t Not Installed", sep='')

	try:
		out_bytes = subprocess.check_output(['bundle', '-v'])
		out_text = out_bytes.decode('UTF-8')
		out_text = out_text.split(' ')
		bVersion = out_text[2]
		print("\t[*] Bundle Version:\t", bVersion, sep='')
	except:
		print("\t[*] Bundle Version:\t Not Installed", sep='')
	return

# TODO: Catch exceptions when Git fails update because local file was modified
def do_git_apps(install_path, tools_dict):
	"""
	Usage: do_git_apps(base_install_dir, dict_appname_github_url)
	Will process a dictionary of tools
	If dict value is a tuple, it will process post-clone installation
	"""
	for app, url in tools_dict.iteritems():
		install_script = False	# Need to initialize for function to work
		print("\n[*] Checking Repository: {}".format(app))
		app_path = os.path.join(install_path, app)
		# If the app already exists, just update it
		if os.path.isdir(app_path):
			os.chdir(app_path)
			try:
				subprocess.call('git pull', shell=True)
				time.sleep(tdelay)
			except:
				print("[-] Error updating Git Repo: {}".format(app))
		else:
			# App does not exist, we can set it up right now
			print("[-] This path does not exist: {}".format(app_path))
			createMe = input("Setup this Git Clone now? [y,N]: ")
			if(createMe == 'y'):
				make_dirs(app_path)
				os.chdir(install_path)
				# Check if the dict value is a tuple, indicating special tools
				if type(tools_dict[app]) is tuple:
					url2, install_script = url
				else:
					# Normal git, so just standardize the variable being used
					url2 = url
				subprocess.call('git clone ' + str(url2) + '.git ' + str(app), shell=True)

				# Now process all special git apps, which require additional installation
				if install_script:
					run_helper_script(os.path.join(app_path, install_script))
		# Check and update the Veil framework
		if app == 'veil':
			if UPDATE_VEIL:
				run_helper_script(os.path.join(app_path, 'update.sh'))
	return

def run_helper_script(script_path):
	"""
	This function aids installs and updates by running developers'
	locally crafted scripts. No sense in redoing their hard efforts.
	However, this also introduces a security concern if said script is malicious
	"""
	if not os.path.isfile(script_path):
		print("[-] Script path is invalid, skipping script execution")
	if not os.access(script_path, os.X_OK):
		try:
			os.chmod(script_path, 0o755)
		except:
			print("[-] Unable to modify permissions on script file, aborting")
			return False
	os.system('clear')
	subprocess.call(script_path, shell=True)
	return True

	
def setup_chrome():
	"""
	Google Chrome has an odd setup path within Kali and hates running as root
	"""
	# Credit to: http://www.blackmoreops.com/2013/12/01/install-google-chrome-kali-linux-part-2/
	chrome_repo = 'deb http://dl.google.com/linux/chrome/deb/ stable main'

	# Google Chrome creates its own sources file, let's check for it first
	if os.path.isfile('etc/apt/sources.list.d/google-chrome.list'):
		found = True
	else:
		f = open('/etc/apt/sources.list', 'r')
		found = False
		for line in f.readlines():
			if line.startswith("#"):
				continue
			if line.strip() == chrome_repo:
				found = True
		f.close()
	if not found:
		try:
			f = open('/etc/apt/sources.list', 'wa')
			f.write(chrome_repo)
			f.close()
		except OSError:
			print("[-] Error trying to write to sources.list file. Installation aborted.")
			return
	#os.system('echo "deb http://dl.google.com/linux/chrome/deb/ stable main" >> /etc/apt/sources.list.d/google-chrome.list')

	subprocess.call('apt-get -qq update', shell=True)
	subprocess.call('apt-get install -y google-chrome-stable', shell=True)

	inputfile = open('/opt/google/chrome/google-chrome', 'r')
	outputfile = open('/tmp/google-chrome', 'w')
	# TODO: This part not done
	return


def maint_tasks():
	if gitcolorize == True:
		os.system("git config --global color.ui auto")
	if fixportmapper == True:
		os.system("update-rc.d rpcbind defaults")
		time.sleep(tdelay)
		os.system("update-rc.d rpcbind enable")
	return

def do_app_updates():
	print("[*] Custom App Updates")
	return

# ------------------------
#          MAIN
# ------------------------
def main():
	maint_tasks()
	core_update()
	print("[+] Kali core update is complete. Listing utility bundle versions below:")
	get_versions()
	print("[+] Now updating Github cloned repositories...")
	do_git_apps(GIT_BASE_DIR, normal_git_tools)
	do_git_apps('/opt', special_git_tools)
	# Check for Google Chrome
	if DO_CHROME:
		setup_chrome()

	do_app_updates()
	print("\n[+] Kali Updater is now complete. Goodbye!")
	return

if __name__ == '__main__':
	main()



kaliupdater.py
#!/usr/bin/env python
#
#   Target Release:     Python 2.x
#   Version:            1.5
#   Updated:            01/04/2015
#   Created by:         cashiuus@gmail.com
#
# Functionality:
#   1. Add Bleeding Edge Repository
#   2. Update & Upgrade all Kali apps using 'apt-get'
#   3. Update defined list of Git Clones; if nonexistant you can add them
# ====================================================================

from __future__ import print_function
import os
import shutil
import subprocess
import sys
import tarfile
import time

# TODO: There is a better way to import external variables
try:
    # Customized version of the "default.py" file
    from settings import *
except:
    # If "settings.py" doesn't exist, import default instead
    from default import *

# ----------------------------- #
#    Set configurable settings  #
# ----------------------------- #
tdelay = 2                  # Delay script for network latency
G  = '\033[32;1m'           # green
R  = '\033[31m'             # red
O  = '\033[33m'             # orange
W  = '\033[0m'              # white (normal)

#    List your existing GIT CLONES here
#    TODO: Get existing git clones automatically; only known way is too slow
#           If I can find them in background during core update, would be ideal
normal_git_tools = {
    'artillery': 'https://github.com/trustedsec/artillery',
    'creds': 'https://github.com/DanMcInerney/creds.py',
    'lair': 'https://github.com/fishnetsecurity/Lair',
    'powersploit': 'https://github.com/mattifestation/PowerSploit',
    'seclists': 'https://github.com/danielmiessler/SecLists',
    'vfeed': 'https://github.com/toolswatch/vFeed',
}

special_git_tools = {
    # These projects have specific directory requirements
    # Key=project_local_name
    #   Value= (nested dict)
    #       Key= install|url|script
    #           Value= path|url|script_basename
    'smbexec':
    {
        'install': '/opt',
        'url': 'https://github.com/pentestgeek/smbexec',
        'script': 'install.sh',
    },
    'veil':
    {
        'install': GIT_BASE_DIRS[0],
        'url': 'https://github.com/Veil-Framework/Veil',
        'script': 'update.sh',
    },
}

# Setup Python 2 & 3 'raw_input' compatibility
#
try:
    input = raw_input
except NameError:
    pass


# ----------------------------- #
#      UTILITY FUNCTIONS        #
# ----------------------------- #
def printer(msg, color=O):
    if color == O and DO_DEBUG:
        print("{}[DEBUG] {!s}{}".format(color, msg, W))
    elif color != O:
        print("{}{!s}{}".format(color, msg, W))
    return


# Check - Root user
# TODO: If not root, run with sudo
def root_check():
    if not (os.geteuid() == 0):
        printer("[-] Not currently root user. Please fix.", color=R)
        exit(1)
    return


def make_dirs(path):
    if not os.path.isdir(path):
        os.makedirs(path)
    return


# ----------------------------- #
#              BEGIN            #
# ----------------------------- #
#    Update Kali core distro using Aptitude
def core_update():
    print("[*] Now updating Kali and Packages...")
    try:
        subprocess.call("apt-get -qq update && apt-get -qq -y dist-upgrade && apt-get -y autoclean", shell=True)
        subprocess.call("updatedb", shell=False)
        print("[*] Successfully updated Kali...moving along...")
    except:
        printer("[-] Error attempting to update Kali. Please try again later.", color=R)

    time.sleep(tdelay)
    return


# -----------------------------
# Checking APT Repository List file
#
def bleeding_repo_check():
    bleeding = 'deb http://repo.kali.org/kali kali-bleeding-edge main'
    repofile = open('/etc/apt/sources.list', 'r')

    for line in repofile.readlines():
        if bleeding == line.rstrip():
            check = 1
            break
        else:
            check = 0
    repofile.close()

    if check == 0:
        addrepo = input("\n[+] Would you like to add the Kali Bleeding Edge Repo? [y,N]: ")
        if (addrepo == 'y'):
            repofile = '/etc/apt/sources.list'
    return


# TODO: Add code to compare current versions with those found to indicate update available
# Get installed version of supporting programming applications with dpkg
def get_versions():
    pversion = str(sys.version_info[0]) + '.' + str(sys.version_info[1]) + '.' + str(sys.version_info[2])
    print("\t{}[*] Python Version:{}\t{}".format(G, W, pversion))

    try:
        out_bytes = subprocess.check_output(['ruby', '-v'])
        out_text = out_bytes.decode('UTF-8')
        rversion = out_text.split(' ')[1]
        print("\t{}[*] Ruby Version:{}\t{!s}".format(G, W, rversion))
    except:
        printer("\t[*] Ruby Version:\t Not Installed", color=R)

    try:
        out_bytes = subprocess.check_output(['gem', '-v'])
        gversion = out_bytes.decode('UTF-8')
        gversion = gversion.rstrip()
        print("\t{}[*] Gem Version:{}\t{!s}".format(G, W, gversion))
    except:
        printer("\t[*] Gem Version:\t Not Installed", color=R)

    try:
        out_bytes = subprocess.check_output(['bundle', '-v'])
        out_text = out_bytes.decode('UTF-8')
        bversion = out_text.split(' ')[2]
        print("\t{}[*] Bundle Version:{}\t{!s}".format(G, W, bversion))
    except:
        printer("\t[-] Bundle Version:\t Not Installed", color=R)
    return


def git_new(app_path, install_path, url, app_script=None):
    # App does not exist, we can set it up right now
    print("[-] This path does not exist: {}".format(app_path))
    answer = input("Setup this Git Clone now? [y,N]: ")
    if(answer == 'y'):
        make_dirs(app_path)
        # Change dir to the install_path, not app_path, so that clone creates dir in the right place
        os.chdir(install_path)

        # Clone the app into a new folder called $app
        subprocess.call('git clone ' + str(url) + '.git ' + str(app), shell=True)

        # Now process all special git apps, which require additional installation
        if app_script:
            run_helper_script(os.path.join(app_path, app_script))
    return


# TODO: Catch exceptions when Git fails update because local file was modified
def do_git_apps(path_list, tools_dict, special_tools):
    """
    Usage: do_git_apps(list(base_install_dir), dict(appname_github_url))
    Will process a dictionary of tools
    If dict value is a dict, it will process post-clone installation
    """
    fnull = open(os.devnull, 'w')

    def git_update(git_path):
        if os.path.isdir(git_path):
            os.chdir(git_path)
            try:
                subprocess.call('git pull', shell=True)
                time.sleep(tdelay)
            except:
                printer("[-] Error updating Git Repo: {}".format(git_path))
                # TODO: Maybe in future ask if we want to reset this broken repo, not yet
            return True
        else:
            return False

    def git_owner(ap):
        # Get the owner for existing Git repos
        # the .git/config file has a line that startswith url that contains remote url
        with open(os.path.join(ap, '.git', 'config'), 'r') as fgit:
            for line in fgit.readlines():
                if line.strip().startswith('url'):
                    owner_string = line.strip().split(' ')[2]
        return owner_string


    # Pre-processing of existing GIT repositories
    my_apps = []
    for item in path_list:
        # if we have a folder that also has a .git folder inside it, we want to update it
        # list - full paths to app
        # TODO: Maybe a better approach is to find them and ask to update the repo list with existing ones
        my_apps.extend([os.path.join(item, x) for x in os.listdir(item) if os.path.isdir(os.path.join(item, x, '.git'))])

    for app, details in tools_dict.iteritems():
        printer("\n[*] Checking Repository: {}".format(app), color=G)

        # Normal git, so just standardize the variable being used
        url = details
        # New repo clones are always installed in first directory path
        install_path = path_list[0]
        app_path = os.path.join(install_path, app)
        print("[*] Application Path: {}".format(app_path))
        printer("[*] Git Owner: {}".format(git_owner(app_path)), color=G)
        # Avoid redundancy, remove apps from list if we're checking it already
        if app_path in my_apps:
            my_apps.remove(app_path)
            
        if not git_update(app_path):
            git_new(app_path, install_path, url, app_script=app_script)

    for app, details in special_tools.iteritems():
        printer("\n[*] Checking Repository: {}".format(app), color=G)
        app_script = False      # Initialize toggle for function to work
        # What type of mapping are we dealing with
        # Check if the dict value is a dict, indicating special tools
        printer("Special Tools Mapping: {}".format(app))
        install_path = special_tools[app]['install']
        url = special_tools[app]['url']
        app_script = special_tools[app]['script']
        app_path = os.path.join(install_path, app)
        print("[*] Application Path: {}".format(app_path))
        
        # Avoid redundancy, remove apps from list if we're checking it already
        if app_path in my_apps:
            my_apps.remove(app_path)
            
        if not git_update(app_path):
            git_new(app_path, install_path, url, app_script=app_script)
    # Now update my existing git repoos
    for i in my_apps:
        printer("\n[*] Updating Repository: {}".format(i.split('/')[-1]), color=G)
        git_update(i)


def run_helper_script(script_path):
    """
    This function aids installs and updates by running developers'
    locally crafted scripts. No sense in redoing their hard efforts.
    However, this also introduces a security concern if said script is malicious
    """
    if not os.path.isfile(script_path):
        printer("[-] Script path is invalid, skipping script execution", color=R)
        return False
    if not os.access(script_path, os.X_OK):
        try:
            os.chmod(script_path, 0o755)
        except:
            printer("[-] Unable to modify permissions on script file, aborting", color=R)
            return False
    os.system('clear')
    subprocess.call(script_path, shell=True)
    return True


def setup_chrome():
    """
    Google Chrome has an odd setup path within Kali and hates running as root
    """
    # Credit to: http://www.blackmoreops.com/2013/12/01/install-google-chrome-kali-linux-part-2/
    chrome_repo = 'deb http://dl.google.com/linux/chrome/deb/ stable main'

    # Google Chrome creates its own sources file, let's check for it first
    if os.path.isfile('/etc/apt/sources.list.d/google-chrome.list'):
        found = True
    else:
        f = open('/etc/apt/sources.list', 'r')
        found = False
        for line in f.readlines():
            if line.startswith("#"):
                continue
            if line.strip() == chrome_repo:
                found = True
        f.close()
    if not found:
        try:
            f = open('/etc/apt/sources.list', 'a')
            f.write(chrome_repo)
            f.close()
        except OSError:
            printer("[-] Error trying to write to sources.list file. Installation aborted.", color=R)
            return

    # Check if Chrome is already installed in its default location
    if not os.path.isfile('/opt/google/chrome/google-chrome'):
        subprocess.call('apt-get install -qq -y google-chrome-stable', shell=True)

    # Check Chrome's config file and fix if it is broken for root use
    inputfile = open('/opt/google/chrome/google-chrome', 'r')
    contents = inputfile.readlines()
    inputfile.close()
    if contents[-1].strip() == '"$@"':
        # The default config file is back and we must replace it
        outputfile = open('/tmp/google-chrome', 'w')
        for line in contents[:-2]:
            outputfile.write(line)
        chromefix = 'exec -a "$0" "$HERE/chrome" "$@" --user-data-dir "$PROFILE_DIRECTORY_FLAG"\n'
        outputfile.write(chromefix)
        outputfile.close()
        # Move the new file to overwrite the old
        try:
            shutil.move('/tmp/google-chrome', '/opt/google/chrome/google-chrome')
            run_helper_script('/opt/google/chrome/google-chrome')
        except Exception as e:
            printer("Error moving Google chrome config file: {}".format(e))
            pass
    else:
        print("[*] Google chrome config file appears to be fine, moving along")
    return


def maint_tasks():
    if gitcolorize is True:
        subprocess.check_output(['git', 'config', '--global', 'color.ui', 'auto'])
    if fixportmapper is True:
        subprocess.call("update-rc.d rpcbind defaults", shell=True)
        time.sleep(tdelay)
        subprocess.call("update-rc.d rpcbind enable", shell=True)
    return


def file_filter(f):
    if f in EXCLUDE_FILES:
        return True
    else:
        return False


def backup_files(files, dest):
    if not os.path.isdir(BACKUP_PATH):
        try:
            os.makedirs(BACKUP_PATH, mode=0o711)
        except:
            printer("[-] Error creating backup folders", color=R)
            return False

    # Create the compressed archive the files will be sent to
    zname = BACKUP_PATH + os.sep + 'daily-' + time.strftime('%Y%m%d') + '.tar.gz'
    if os.path.exists(zname):
        response = input("[-] The backup destination file already exists, overwrite? [y, N]: ")
        if response != 'y':
            return False
    z = tarfile.open(zname, mode='w:gz')

    # ------ dotfiles -----------
    for f in os.listdir(usr):
        fpath = os.path.join(usr, f)
        if f.startswith('.') and not os.path.isdir(fpath):
            if f not in EXCLUDE_FILES:
                # Add to archive if it is not in the excluded files list
                z.add(fpath)

    # --------- backup files ----------
    for fpath in BACKUP_FILES:
        try:
            # Add files unless they are listed in the exclude settings
            z.add(fpath, filter=lambda x: None if x.name in EXCLUDE_FILES else x)
        except Exception as e:
            printer("Error adding backup file: {}".format(e), color=R)
    # Close archive file
    z.close()
    return True


# ------------------------
#          MAIN
# ------------------------
def main():
    maint_tasks()
    if DO_CHROME:
        # Do chrome first, this way we can run update just once
        setup_chrome()
    core_update()
    print("[*] Kali core update is complete. Listing utility bundle versions below:")
    get_versions()
    if DO_GIT_REPOS:
        print("[*] Now updating Github cloned repositories...")
        do_git_apps(GIT_BASE_DIRS, normal_git_tools, special_git_tools)

    if DO_BACKUPS:
        if backup_files(BACKUP_FILES, BACKUP_PATH) is True:
            print("[*] Backups successfully saved to: {}".format(BACKUP_PATH))
        else:
            printer("[-] Backups failed to complete", color=R)

    printer("\n[*] Kali Updater is now complete. Goodbye!\n", color=G)
    return


if __name__ == '__main__':
    main()
